src/libsyntax/parse/parser.rs

```diff
@@ -5114,7 +5120,39 @@ impl<'a> Parser<'a> {
                     if has_parens {
                         self.expect(&token::CloseDelim(token::Paren))?;
                     }
-                    let poly_trait = PolyTraitRef::new(lifetime_defs, path, lo.to(self.prev_span));
+                    let poly_span = lo.to(self.prev_span);
+                    if is_negative {
+                        let mut err = self
+                            .struct_span_err(poly_span,
+                                             "negative trait bounds are not supported");
+                        if colon_span.is_some() {
+                            err.span_suggestion_with_applicability(
+                                last_plus_span.or_else(|| {
+                                    has_removed_colon = true;
+                                    colon_span
+                                }).unwrap().to(poly_span),
+                                "remove this trait bound",
+                                "".into(),
+                                Applicability::MachineApplicable);
+                        } else {
+                            err.help("remove this trait bound");
+                        }
+                        err.emit();
+                    } else if has_removed_colon {
+                        has_removed_colon = false;
+                        self.diagnostic()
+                            .struct_span_warn(
+                                poly_span,
+                                "this trait bound will become the first one")
+                            .span_suggestion_short_with_applicability(
+                                last_plus_span.unwrap(),
+                                "replace this `+` with `:`",
+                                ":".into(),
+                                Applicability::MachineApplicable)
+                            .emit();
+                    }
```

Comment:
As neat as the output is, it's unadvisable to have multiple suggestions that need to be applied together for them to be correct in separate diagnostics. A more appropriate solution would be to batch all the errors to be emitted (or at least the elements needed to create diagnostics) and emit a single, unambiguous diagnostic with the full suggestion. This would have the nice side effect of also dealing with `Tr5` with a single diagnostic, instead of one per negative trait bound.

Do you feel this is reasonable?